daemon/cmd/daemon_main.go

```diff
@@ -1436,6 +1437,22 @@ func (d *Daemon) instantiateAPI() *restapi.CiliumAPI {
 	return restAPI
 }
 
+func checkifKernelSockmapOptionsConfigured() {
+	if option.Config.SockopsEnable {
+		found := false
+		if h := probes.NewProbeManager().GetHelpers("sock_ops"); h != nil {
```

Comment:
I don't think this is enough; it looks like at least the [`msg_verdict` attach type](https://github.com/torvalds/linux/commit/4f738adba30a7cfc006f605707e7aee847ffefa0) and [the sockhash map type](https://github.com/torvalds/linux/commit/81110384441a59cff47430f20f049e69b98c17f4) were introduced after [the `sock_op` program type](https://github.com/torvalds/linux/commit/40304b2a1567fecc321f640ee4239556dd0f3ee0). Could you double check? I think we need to verify support for all of these.

Reply:
> I don't think this is enough; it looks like at least the [`msg_verdict` attach type](https://github.com/torvalds/linux/commit/4f738adba30a7cfc006f605707e7aee847ffefa0) and [the sockhash map type](https://github.com/torvalds/linux/commit/81110384441a59cff47430f20f049e69b98c17f4) were introduced after [the `sock_op` program type](https://github.com/torvalds/linux/commit/40304b2a1567fecc321f640ee4239556dd0f3ee0). Could you double check? I think we need to verify support for all of these.

I am not sure exactly what should be the prog and map type. Should it be 'sk_msg' instead of 'sock_ops' or 'sk_msg_redirect' and for the maps should we need to check for 'bpf_sock_map_hash_update'?

Reply:
I'm expecting the sockhash map is accessed through the usual map helpers so you'll need to check the sockhash map itself is supported.

It looks like `bpftool feature` doesn't expose attach type support yet, so maybe checking for the `sk_msg` program type instead would be enough for the time being. It was introduced in the same commit as the `msg_verdict` attach type.

Reply:
@pchaigno For some reason still even with 'sk_msg' the bpftool does not return any helper functions. Is there anything I am missing?

Reply:
@soumynathan It does, but `bpf_sock_hash_update` is not part of those helpers:
```
$ sudo bpftool feature probe macros | grep sk_msg | grep "1$" | wc -l
20
$ sudo bpftool feature probe macros | grep sk_msg | grep bpf_sock_hash_update
#define BPF__PROG_TYPE_sk_msg__HELPER_bpf_sock_hash_update 0
```
I'm not sure why you're checking support for the `bpf_sock_hash_update` helper though. Wouldn't checking support for `sk_msg` and `sock_op` program types be enough?

Reply:
Ok, I think with the latest kernel, I am able to list out the 'bpf' helper functions for the 'sk_msg'. What are the functions that we should be checking, probably
```
                    bpf_msg_pop_data:{}
                    bpf_msg_pull_data:{}
                    bpf_msg_push_data:{}
                    bpf_msg_redirect_hash:{}
                    bpf_msg_redirect_map:{}

```
```
Apr 16 01:38:46 k8s1 cilium-agent[17595]: level=warning msg="BPF Sock ops.%smap[bpf_get_numa_node_id:{}
                    bpf_get_prandom_u32:{}
                    bpf_get_smp_processor_id:{}
                    bpf_jiffies64:{}
                    bpf_ktime_get_ns:{}
                    bpf_map_delete_elem:{}
                    bpf_map_lookup_elem:{}
                    bpf_map_peek_elem:{}
                    bpf_map_pop_elem:{}
                    bpf_map_push_elem:{}
                    bpf_map_update_elem:{}
                    bpf_msg_apply_bytes:{}
                    bpf_msg_cork_bytes:{}
                    bpf_msg_pop_data:{}
                    bpf_msg_pull_data:{}
                    bpf_msg_push_data:{}
                    bpf_msg_redirect_hash:{}
                    bpf_msg_redirect_map:{}
                    bpf_spin_lock:{}
                    bpf_spin_unlock:{}
                    bpf_tail_call:{}]

Reply:
Our `sk_msg` program seems to only be using base helpers and `msg_redirect_hash()`. The latter was introduced at the same time as the sockhash map type so I don't think we need to check for any helper here.

Reply:
> I'm expecting the sockhash map is accessed through the usual map helpers so you'll need to check the sockhash map itself is supported.
> 
> It looks like `bpftool feature` doesn't expose attach type support yet, so maybe checking for the `sk_msg` program type instead would be enough for the time being. It was introduced in the same commit as the `msg_verdict` attach type.

So we are currently collecting the helper functions for 'sk_msg'? So just checking for 'sk_msg' program types would be enough?

Reply:
@soumynathan I think you need to check support for:
1. the `sk_msg` program type;
0. the `sock_op` program type;
0. `sockhash` map type.

Reply:
Thanks for the clarification. Will address your review comments.